### PR TITLE
Add _WD_LocateElement

### PR DESCRIPTION
**Pull request by mlipok**
### Proposed changes

There should be easy way to find element location in the frame list returned by `_WD_FrameList`

### Checklist

- [x] I have read and noticed the [CODE OF CONDUCT](https://github.com/Danp2/au3WebDriver/blob/master/docs/CODE_OF_CONDUCT.md) document
- [x] I have read and noticed the [CONTRIBUTING](https://github.com/Danp2/au3WebDriver/blob/master/docs/CONTRIBUTING.md) document
- [ ] I have added necessary documentation or screenshots (if appropriate)

### Types of changes

Please check `x` the type of change your PR introduces:

- [ ] Bugfix (change which fixes an issue)
- [x] Feature (change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (functional, structural)
- [ ] Documentation content changes
- [ ] Other (please describe)

### What is the current behavior?
No easy way to find element location in the frame list because `_WD_FindElement` finds element only in current location (  main document or desired frame selected by `_WD_FrameEnter`  )

### What is the new behavior?
It shoul be easie with `_WD_ElementExist`

### Additional context
This was initially discussed here: 
https://github.com/Danp2/au3WebDriver/pull/362#issuecomment-1207494392

and here
https://github.com/Danp2/au3WebDriver/pull/362#discussion_r944392108

### System under test
not related

### CODE REVIEW

**Comment by mlipok** [comment]
@Danp2 
Before I proceed with further work on this functionality, please check if you will initially accept it.

**Comment by mlipok** [comment]
Please check it with modified [wd_demo.au3](https://github.com/Danp2/au3WebDriver/pull/386/commits/b7712f77aa1899ff72192a76828a253be8d5368f)

**Comment by Danp2** [comment]
> Please check it with modified [wd_demo.au3](https://github.com/Danp2/au3WebDriver/pull/386/commits/b7712f77aa1899ff72192a76828a253be8d5368f)

I did. Unfortunately, it doesn't work for me. This is what I get in the logs --

```
_WD_ElementExist ==> Success [0] : Parameters:   Strategy=css selector   Selector=div.trytopnav   ShadowRoot=Default    : LocationOfElement= null
wd_demo.au3: (583) : $sLocationOfElement1=null
<snip>
_WD_ElementExist ==> Success [0] : Parameters:   Strategy=css selector   Selector=div.trytopnav   ShadowRoot=Default    : LocationOfElement= null
wd_demo.au3: (606) : $sLocationOfElement2=null
```

**Comment by mlipok** [comment]
And this are correct result.
If you want other example provide another website where desired element is located in inner frame.

**Comment by Danp2** [comment]
OIC. I misinterpreted the results by assuming that "null" was incorrect.

**Comment by mlipok** [comment]
> OIC. I misinterpreted the results by assuming that "null" was incorrect.

So I assume that now you understand that null is correct location for this element, as it is located on top window

**Comment by Danp2** [comment]
@mlipok Some feedback on this PR --

- I'm still trying to understand the "Use Case" for this function. 
- I find the function name misleading.  Have you considered any other names (ie: something like `_WD_ElementContext`)?
- How do you plan to handle multiple matching elements? Hidden / disabled elements?

**Comment by mlipok** [comment]
> * I'm still trying to understand the "Use Case" for this function.

When non advance user is using HTML dom to find element by XPATH or by CSSQuerySelector they do not bother about frames as because browser find them just a way.
So user is using `_WD_FindElement` which as you can expect in case of frames do not find the desired element, because frame was not changed.

You can even teach user how to use browser to check if there are frames. But how would you like to pass the way of identifying a particular frame to the format required by `_WD_FrameEnter` ?

If user say my XPATH works in browser but not with `_WD_FindElement` then you should say use the same XPATH with `_WD_ElementExist` / `_WD_ElementeContext` and result use with `_WD_FrameEnter` before you try to use `_WD_FindElement`

**Comment by mlipok** [comment]
btw.
Even I have sometimes problem with localizing the element in HTML DOM context .
Of course I find them but it should be easier, and for this reason I create this function.
As you can locate them easily on browser but this is not always easy to get the rigth path like `null/2/0/1`

**Comment by mlipok** [comment]
> * How do you plan to handle multiple matching elements? Hidden / disabled elements?

Multiple matching is good point. And will be added soon in this PR.
Firstly I want be sure that current state is fine, and I can do next step = `multiple element support`

Hidden/Disabled that is not our concern. We only finding element locations. I can compare it to `_WD_ElementFind` which also to not bother about Hidden/Disabled status.

**Comment by mlipok** [comment]
> * I find the function name misleading.  Have you considered any other names (ie: something like `_WD_ElementContext`)?

My observation:

- The au3WebDriver UDF is using `Frame` in fucntion names not `Context`.
- We are planing to have multiple `Elements` support
- specific `XPATH` or `CSSQuerySelector` can locate multiple `Elements` on multiple `Frames`

My new proposals for function name are: `_WD_GetElementsFramesLocations` or `_WD_ElementsFramesLocations` or `_WD_ElementsFrames`

**Comment by mlipok** [comment]
> * I'm still trying to understand the "Use Case" for this function.

try to use:
```autoit
_WD_Navigate($sSession, "https://www.tutorialspoint.com/html/html_frames.htm#")
_WD_ELementExist($sSession, $_WD_LOCATOR_ByCSSSelector, "li.nav-item[data-bs-original-title='Home Page'] a.nav-link[href='https://www.tutorialspoint.com/index.htm']")
```

**Comment by Danp2** [comment]
>  As you can locate them easily on browser but this is not always easy to get the rigth path like null/2/0/1

It's been a while since I've had to work with frames, but IIRC I always located the frame first with _WD_FindElement and then passed the Element ID to _WD_Window / _WD_FrameEnter.

> My new proposals for function name are: _WD_GetElementsFramesLocations or _WD_ElementsFramesLocations or _WD_ElementsFrames

Only the last one has any chance of approval. 😉 Another for consideration is _WD_FindElementFrames.

**Comment by mlipok** [comment]
`_WD_FindElementFrames` is the best.

**Comment by mlipok** [comment]
> It's been a while since I've had to work with frames, but IIRC I always located the frame first with _WD_FindElement and then passed the Element ID to _WD_Window / _WD_FrameEnter.

I had today an issue related to frames which I solve using `_WD_FrameList` - I just observe when relative path was changed.

**Comment by mlipok** [comment]
> * I'm still trying to understand the "Use Case" for this function.

So is it now clear to you what is the purpose of this function ?

**Comment by Danp2** [comment]
> So is it now clear to you what is the purpose of this function ?

I tried running your revised demo, but it didn't complete successfully --

```
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/0 Information: Error occured on "null/0" level when trying to check atributes child frames #0
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=True Information:  Was not able to check / back to "calling frame"
! ---> @error=10  @extended=0 : Example 7
_WD_ElementExist ==> General Error [1] : Parameters:   Strategy=css selector   Selector=li.nav-item[data-bs-original-title='Home Page'] a.nav-link[href='https://www.tutorialspoint.com/index.htm']   ShadowRoot=Default > Was not able to check / back to "calling frame" : StartLocatkon=    : LocationOfElement= 
wd_demo.au3: (711) : $sLocationOfElement3=
```

The error seems to originate on this line -- https://github.com/Danp2/au3WebDriver/blob/0eb6cc9cdb97e42219ed76485885c46b16ed5209/wd_helper.au3#L828

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/48e7da9c-3a00-4644-b871-ba2fa15123ea/execute/sync; Data={"script":"return document.querySelectorAll('iframe')[0].outerHTML;", "args":[]}
__WD_Post ==> Javascript Exception [21] : HTTP status = 500 ResponseText={"value":{"error":"javascript error","message":"TypeError: document.querySelectorAll(...)[0] is undefined","stacktrace":"@https://www.tutorialspoint.com/html/src/creating_frames.htm:2:23\n@https://www.tutorialspoint.com/html/src/creating_frames.htm:3:8\n"}}
_WD_ExecuteScript ==> Javascript Exception [21] : Error occurred when trying to ExecuteScript
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=True Information:  Was not able to check / back to "calling frame"
! ---> @error=10  @extended=0 : Example 7
```

File: wd_helper.au3 [hunk 1]
```diff
@@ -1155,6 +1155,85 @@ Func _WD_jQuerify($sSession, $sjQueryFile = Default, $iTimeout = Default)
 	Return SetError(__WD_Error($sFuncName, $iErr, $sParameters))
 EndFunc   ;==>_WD_jQuerify
 
+; #FUNCTION# ====================================================================================================================
+; Name ..........: _WD_FindElementFrames
+; Description ...: Check if element exist in all documents (including frames) and return their location
```
Comment (mlipok) [inline wd_helper.au3:1160]: Please review Description and suplement heplper function list here:
https://github.com/Danp2/au3WebDriver/blob/master/README.md#documentation

**Comment by Danp2** [comment]
> So is it now clear to you what is the purpose of this function ?

I understand what it does. It just doesn't feel "natural" to me since I've never located a frame in that manner.

File: wd_helper.au3 (hunk 1 above)
Comment (Danp2) [reply wd_helper.au3:1160]: The description needs work. My first crack at it is "Search the current document (including frames) and return the location of any matching elements". 

Next you'll need to define what "location" means. Perhaps an entry under "Remarks"

**Comment by mlipok** [comment]
> It just doesn't feel "natural" to me since I've never located a frame in that manner.

I am not surprised as the ability to specify a path is a new feature ;)

File: wd_helper.au3 (hunk 1 above)
Comment (mlipok) [reply wd_helper.au3:1160]: `Search the current document (including frames) and return the location of any matching elements`

Any suplement/fix ... if we takes into consideration that matching elements can be found on multiple frames ? I mean list of frames location will be returned as array.

Comment (Danp2) [reply wd_helper.au3]: > Any suplement/fix ... if we takes into consideration that matching elements can be found on multiple frames ? I mean list of frames location will be returned as array.

I don't understand this last comment. Isn't this already covered by `return the location of any matching elements`?

Comment (mlipok) [reply wd_helper.au3]: I re-read it and all is Ok here.

File: README.md [hunk 2]
```diff
@@ -78,6 +78,7 @@ This au3WebDriver UDF (project) allows to interact with any browser that support
 | _WD_ElementOptionSelect | Find and click on an option from a Select element.                              |
 | _WD_ElementSelectAction | Perform action on designated Select element.                                    |
 | _WD_ElementStyle        | Set/Get element style property.                                                 |
+| _WD_FindElementFrames   | Search the current document (including frames) and return locations of matching elements    |
```
Comment (mlipok) [inline README.md:81]: should we also add additionall spaces to reposiotion `|` for other functions ?

File: wd_helper.au3 (hunk 1 above)
Comment (mlipok) [reply wd_helper.au3]: > Next you'll need to define what "location" means. Perhaps an entry under "Remarks"

What you think about my proposed remark:
`Returned location (path like 'null/2/0') can be used with _WD_FrameEnter before _WD_FindElement or _WD_WaitElement will be used.`

File: README.md (hunk 2 above)
Comment (Danp2) [reply README.md:81]: Sure. FWIW, I've used [this site](https://www.tablesgenerator.com/markdown_tables) in the past to quickly layout tables in markdown format.

Comment (mlipok) [reply README.md:81]: Please take a look on: [Helper Functions - table indentation](https://github.com/Danp2/au3WebDriver/pull/386/commits/9c625a1350e862a8c55d005568c2e14c29598cbc)

File: wd_helper.au3 (hunk 1 above)
Comment (Danp2) [reply wd_helper.au3]: I think it is too wordy / confusing. I suggest changing it to "The returned location can be used with _WD_FrameEnter to switch to the target frame."

**Review by Danp2 — COMMENTED** [review]
@mlipok Did you miss this? https://github.com/Danp2/au3WebDriver/pull/386#issuecomment-1227826674

File: wd_helper.au3 (hunk 1 above)
Comment (mlipok) [reply wd_helper.au3]: Please take a look: [Remarks](https://github.com/Danp2/au3WebDriver/pull/386/commits/e0eb4fac253f52affc6da250cdd83930adfc10bf)

**Comment by mlipok** [comment]
> @mlipok Did you miss this? [#386 (comment)](https://github.com/Danp2/au3WebDriver/pull/386#issuecomment-1227826674)

WIP

**Comment by mlipok** [comment]
When I perform test then I have many issues with internet connectivity.
I think my Zyxel UTM thinks my internet activity is suspicious.

**Comment by mlipok** [comment]
> @mlipok Did you miss this? [#386 (comment)](https://github.com/Danp2/au3WebDriver/pull/386#issuecomment-1227826674)

do you still have this?
If you add 10 seconds of waiting after navigation (to be sure that everything is normal), does the problem stop or is it still happening?

**Comment by Danp2** [comment]
> do you still have this?

Yes

> If you add 10 seconds of waiting after navigation (to be sure that everything is completed), does the problem stop or is it still happening?

Still occurs

**Comment by mlipok** [comment]
Analyzing.

**Comment by mlipok** [comment]
I have already done this function but there is too many conflict.
I also have it developed on currently WIP (not jet merged PR's)
I will back with new PR when ASAP.
Temporary closing this PR.
